gw_eccentricity/eccDefinition.py

```diff
@@ -1184,6 +1228,180 @@ def interp_extrema(self, extrema_type="pericenters"):
+        # value is found by doing some iterations.
+        if not (self.rational_fit_kwargs["num_degree"] and self.rational_fit_kwargs["denom_degree"]):
+            self.rational_fit_kwargs["num_degree"], self.rational_fit_kwargs["denom_degree"] \
+                = self.get_approximate_degree_for_rational_fit()
+
+        rat_fit = self.get_rat_fit(x, y)
+        t = np.arange(x[0], x[-1], self.t[1] - self.t[0])
+        # save the degrees for checks at each step of iterations for finding the optimal degrees
+        old_num_degree = self.rational_fit_kwargs["num_degree"]
+        old_denom_degree = self.rational_fit_kwargs["denom_degree"]
+
+        # use a flag to know whether the initial degrees produced monotonic
+        # time derivative of the fits or it was required to be reduced.
+        degrees_were_reduced = False
+
+        # Check for nonmonotonicity and lower degrees if needed
+        while self.check_if_domega_dt_is_nonmonotonic(t, rat_fit(t), 1.0, description):
+            degrees_were_reduced = True  # Mark that reduction occurred
+            if self.rational_fit_kwargs["num_degree"] > 1:
+                self.rational_fit_kwargs["num_degree"] -= 1
+            if self.rational_fit_kwargs["denom_degree"] > 1:
+                self.rational_fit_kwargs["denom_degree"] -= 1
+            if self.rational_fit_kwargs["num_degree"] == 1 and self.rational_fit_kwargs["denom_degree"] == 1:
+                debug_message("Both numerator and denominator degrees are equal to 1 "
+                              "and cannot be lowered further.",
+                              debug_level=self.debug_level, important=False)
+                break
+
+            debug_message(f"Lowering degrees to num_degree={self.rational_fit_kwargs['num_degree']}, "
+                          f"denom_degree={self.rational_fit_kwargs['denom_degree']} and retrying.",
+                          debug_level=self.debug_level, important=False)
+
+            rat_fit = self.get_rat_fit(x, y)
+        # If no degrees were reduced, try increasing the degree for better fit
+        if not degrees_were_reduced:
+            last_monotonic_rat_fit = rat_fit  # Track last monotonic fit
+            last_monotonic_num_degree = old_num_degree
+            last_monotonic_denom_degree = old_denom_degree
+
+            while not self.check_if_domega_dt_is_nonmonotonic(t, rat_fit(t), 1.0, description):
+                # Increase the degrees for both numerator and denominator
+                new_num_degree = self.rational_fit_kwargs["num_degree"] + 1
+                new_denom_degree = self.rational_fit_kwargs["denom_degree"] + 1
+                self.rational_fit_kwargs["num_degree"], self.rational_fit_kwargs["denom_degree"] \
+                    = new_num_degree, new_denom_degree
+                new_rat_fit = self.get_rat_fit(x, y)
+                if self.check_if_domega_dt_is_nonmonotonic(t, new_rat_fit(t), 1.0, description):
```

Comment:
Looks like this check happens twice for the same fit in some cases? Maybe rework the code so that that doesn't happen?

Reply:
Done